activesupport/lib/active_support/evented_file_update_checker.rb

```diff
@@ -50,32 +51,20 @@ def initialize(files, dirs = {}, &block)
       @block      = block
       @updated    = Concurrent::AtomicBoolean.new(false)
       @lcsp       = @ph.longest_common_subpath(@dirs.keys)
-      @pid        = Process.pid
       @boot_mutex = Mutex.new
 
-      dtw = directories_to_watch
-      @dtw, @missing = dtw.partition(&:exist?)
+      @dtw = directories_to_watch
+      @missing = []
 
       boot!
+      ActiveSupport::ForkTracker.after_fork { boot! }
```

Comment:
The downside to this is that it keeps a reference to the `EventedFileUpdateChecker` instance in `ActiveSupport::ForkTracker`.  Meaning the instance won't be garbage collected if it is ever abandoned.  I can look into a solution involving `WeakRef` if this is a concern.  Or we could design some kind of `stop!` API.

(It's also why `@dtw` is now recomputed in `boot!`.  Otherwise, if an `EventedFileUpdateChecker` instance is created by a test, and the directories it watches are deleted on teardown, that instance's `boot!` would raise when it's re-triggered by [another test's fork](https://github.com/rails/rails/pull/39718/files#diff-0752a2eca9a80ac701287d0e6666234dR48).)

Reply:
https://github.com/rails/rails/blob/a6e98b2267a4470c2de4aa3eb16651c161d98091/actionview/lib/action_view/cache_expiry.rb#L29 looks like one place where we might build a file checker and later abandon it, for what it is worth.

Reply:
That is a good catch! :eyes:  Upon further investigation, I believe that, even without this PR, there is a problem with abandoning checkers due to the way the Listen gem currently works.  I've detailed the problem and proposed some fixes in guard/listen#476, but it hasn't gotten traction yet.  I have a few follow-up Rails PRs in mind, once that issue is addressed.

Reply:
Now that guard/listen#476 has been fixed, I wrote #40100 and #40133 to ensure checkers can be safely abandoned and garbage collected.  I rebased this PR on top of them, and piggy-backed on the finalizer from #40133 to remove the `after_fork` callback from `ActiveSupport::ForkTracker` when the `EventedFileUpdateChecker` instance is garbage collected.